x-pack/plugins/embeddable_enhanced/public/plugin.ts

```diff
@@ -110,12 +141,78 @@ export class EmbeddableEnhancedPlugin
     return true;
   };
 
+  private initializeDynamicActions(
+    uuid: string,
+    getTitle: () => string | undefined,
+    state: DynamicActionsSerializedState
+  ): {
+    dynamicActionsApi: HasDynamicActions;
+    dynamicActionsComparator: EmbeddableStateComparators<DynamicActionsSerializedState>;
+    serializeDynamicActions: () => DynamicActionsSerializedState;
+  } {
+    const dynamicActionsState$ = new BehaviorSubject<DynamicActionsSerializedState['enhancements']>(
+      state.enhancements
+    );
+    const api: DynamicActionStorageApi = {
+      dynamicActionsState$,
+      setDynamicActions: (newState) => dynamicActionsState$.next(newState),
+    };
+    const storage = new ApiActionStorage(uuid, getTitle, api);
+    const dynamicActions = new DynamicActionManager({
+      isCompatible: async (context: unknown) => {
+        return apiHasUniqueId(context) && context.uuid === uuid;
+      },
+      storage,
+      uiActions: this.uiActions!,
+    });
+    this.startDynamicActions(dynamicActions);
+
+    return {
+      dynamicActionsApi: { ...api, enhancements: { dynamicActions } },
+      dynamicActionsComparator: {
+        enhancements: [
+          dynamicActionsState$,
+          api.setDynamicActions,
+          (a, b) => {
+            return deepEqual(a, b);
+          },
+        ],
+      },
+      serializeDynamicActions: () => {
+        return { enhancements: dynamicActionsState$.getValue() };
+      },
+    };
+  }
+
   private enhanceEmbeddableWithDynamicActions<E extends IEmbeddable>(
     embeddable: E
   ): EnhancedEmbeddable<E> {
     const enhancedEmbeddable = embeddable as EnhancedEmbeddable<E>;
 
-    const storage = new EmbeddableActionStorage(embeddable as EmbeddableWithDynamicActions);
+    const dynamicActionsState$ = new BehaviorSubject<DynamicActionsSerializedState['enhancements']>(
+      embeddable.getInput().enhancements as DynamicActionsSerializedState['enhancements']
+    );
+    const api = {
+      dynamicActionsState$,
+      setDynamicActions: (newState: DynamicActionsSerializedState['enhancements']) => {
+        embeddable.updateInput({ enhancements: newState });
+      },
+    };
+
+    embeddable
+      .getInput$()
+      .pipe(
+        distinctUntilChanged(({ enhancements: old }, { enhancements: updated }) =>
+          deepEqual(old, updated)
+        )
+      )
+      .subscribe((input) => {
+        dynamicActionsState$.next(
+          input.enhancements as DynamicActionsSerializedState['enhancements']
+        );
+      });
```

Comment:
We must keep the `dynamicActionsState$` in sync with the input changes - that way, enhanced embeddables will respond correctly to the dashboard being reset (for example)